src/toil/jobStores/ftp_utils.py

```diff
@@ -0,0 +1,214 @@
+        """
+        ftp = self._connect(fn)
+        if ftp:
+            try:
+                cwd = ftp.pwd()
+                ftp.cwd(urlparse(fn).path)
+                ftp.cwd(cwd)
+                return True
+            except ftplib.all_errors:
+                return False
+        return False
+
+    def open(self, fn: str, mode: str) -> IO[bytes]:
+        """
+        Open an FTP url.
+
+        Only supports reading, no write support.
+        :param fn: FTP url
+        :param mode: Mode to open FTP url in
+        :return:
+        """
+        if "r" in mode:
+            host, user, passwd, path = self._parse_url(fn)
+            handle = urlopen("ftp://{}:{}@{}/{}".format(user, passwd, host, path))
+            return cast(IO[bytes], closing(handle))
+        # TODO: support write mode
+        raise Exception("Write mode FTP not implemented")
+
+    def _parse_url(
+        self, url: str
+    ) -> tuple[Optional[str], Optional[str], Optional[str], str]:
+        """
+        Parse an FTP url into hostname, username, password, and path
+        :param url:
+        :return: hostname, username, password, path
+        """
+        parse = urlparse(url)
+        user = parse.username
+        passwd = parse.password
+        host = parse.hostname
+        path = parse.path
+        if parse.scheme == "ftp":
+            if not user and self.netrc:
+                if host is not None:
+                    creds = self.netrc.authenticators(host)
+                    if creds:
+                        user, _, passwd = creds
+        if not user:
+            if host is not None:
+                user, passwd = self._recall_credentials(host)
+                if passwd is None:
+                    passwd = "anonymous@"
+                    if user is None:
+                        user = "anonymous"
+
+        return host, user, passwd, path
+
+    def _connect(self, url: str) -> Optional[ftplib.FTP]:
+        """
+        Connect to an FTP server. Handles authentication.
+        :param url: FTP url
+        :return: FTP object
+        """
+        parse = urlparse(url)
+        if parse.scheme == "ftp":
+            host, user, passwd, _ = self._parse_url(url)
+            if host is None:
+                # there has to be a host
+                return None
+            if (host, user, passwd) in self.cache:
+                if self.cache[(host, user, passwd)].pwd():
+                    return self.cache[(host, user, passwd)]
+            ftp = ftplib.FTP_TLS()
+            ftp.set_debuglevel(1 if logger.isEnabledFor(logging.DEBUG) else 0)
+            ftp.connect(host)
+            env_user = os.getenv("TOIL_FTP_USER")
+            env_passwd = os.getenv("TOIL_FTP_PASSWORD")
+            if env_user:
+                user = env_user
+            if env_passwd:
+                passwd = env_passwd
+            ftp.login(user or "", passwd or "", secure=not self.insecure)
```

Comment:
OK, so according to the obviously reliable https://tedboy.github.io/python_stdlib/_modules/ftplib.html#FTP_TLS.login what `secure` does here is call `auth()` if you haven't already set up SSL/TLS on the connection, before doing the login.

So it looks like unless the user sets `TOIL_FTP_USE_SSL=False` themselves, they're going to get an error here whenever they use an `ftp://` URL, unless the server actually supports SSL/TLS.

One server people might want to use is `ftp://ftp-trace.ncbi.nlm.nih.gov`. I've just checked and it doesn't support FTP over SSL/TLS.

Another thing we don't do is call the [`prot_p()` method](https://docs.python.org/3/library/ftplib.html#ftplib.FTP_TLS.prot_p) on the connection to actually secure the data channel. And later we fall back to `urlopen()` and give it the full FTP URL with username and password, and I don't think that `urlopen()` can use or enforce SSL. We'd have to tinker with `urlopen`'s handler as recommended in https://stackoverflow.com/a/73171311.

So I think that by default we *should* allow unencrypted FTP connections without user intervention to accept them. That's the current behavior.

If we want to change Toil to enforce secure FTP by default, then we also need to enforce security on the actual reads. And we probably would need a way to individually allow particular insecure servers?